python/mlx/nn/layers/transformer.py

```diff
@@ -97,36 +99,71 @@ def create_additive_causal_mask(N: int, dtype: mx.Dtype = mx.float32):
 
 
 class TransformerEncoderLayer(Module):
-    def __init__(self, dims: int, num_heads: int, mlp_dims: Optional[int] = None):
+    def __init__(
+        self,
+        dims: int,
+        num_heads: int,
+        mlp_dims: Optional[int] = None,
+        dropout: float = 0.0,
+        activation: Callable[[Any], Any] = relu,
+        norm_first: bool = False,
+    ):
         super().__init__()
         mlp_dims = mlp_dims or dims * 4
         self.attention = MultiHeadAttention(dims, num_heads)
         self.ln1 = LayerNorm(dims)
         self.ln2 = LayerNorm(dims)
         self.linear1 = Linear(dims, mlp_dims)
         self.linear2 = Linear(mlp_dims, dims)
+        self.dropout1 = Dropout(dropout)
+        self.dropout2 = Dropout(dropout)
+        self.activation = activation
+        self.norm_first = norm_first
 
     def __call__(self, x, mask):
-        y = self.ln1(x)
-        y = self.attention(y, y, y, mask)
-        x = x + y
+        if self.norm_first:
+            y = self.ln1(x)
+            y = self.attention(y, y, y, mask)
+            y = self.dropout1(y)
+            x = x + y
+
+            y = self.ln2(x)
+            y = self.linear1(y)
+            y = self.activation(y)
+            y = self.dropout2(y)
+            y = self.linear2(y)
+            y = x + y
+
+        else:
+            y = self.attention(x, x, x, mask)
+            y = self.dropout1(y)
+            y = self.ln1(x + y)
 
-        y = self.ln2(x)
-        y = self.linear1(y)
-        y = mx.maximum(y, 0)
-        y = self.linear2(y)
-        x = x + y
+            y = self.linear1(y)
+            y = self.activation(y)
+            y = self.dropout2(y)
+            y = self.linear2(y)
+            y = self.ln2(x + y)
```

Comment:
I think the way you have it is long, but not bad because it is very clear. But, if you want the code to be a little shorter you could do something like this:

```
y = self.ln1(x) if self.norm_first else x
y = self.attention(y, y, y, mask)
y = x + self.dropout1(y)
y = y if self.norm_first else self.ln1(y) 

y = self.ln2(y) if self.norm_first else y
y = self.linear(y)
y = self.activation(y)
y = x + self.linear2(y)
y = y if self.norm_first else self.ln1(y) 
```